snapcraft/cli/lifecycle.py

```diff
@@ -151,10 +151,12 @@ def clean(parts, step, **kwargs):
         snapcraft clean
         snapcraft clean my-part --step build
     """
+    if not step:
+        step = 'pull'
+
     project_options = get_project_options(**kwargs)
     container_config = env.get_container_config()
-    if container_config.use_container:
-        step = step or 'pull'
+    if container_config.use_container and step == 'pull' and not parts:
```

Comment:
I don't get this @kalikiana. I might be doing something wrong while testing it, but it seems to me that this still just deletes the container.
The execute method in lxd checks if we are cleaning, and if so, it just deletes the container.
In this pr I see nothing to modify that behaviour.

Reply:
Sorry about that. It looks like the second commit was lost... I re-did the missing piece. So: clean should be called with the container to delete it, in all other cases "normal" lifecycle clean is used so that it works even if the container is broken or already deleted.

Reply:
I'd like to rectify the poor separation of concerns, here. We're going to great lengths to ensure we only ever call `lifecycle.containerbuild('clean')` if `step == 'pull' and not parts`, with no apparent reason. We then proceed to pass the step as well as the parts (that we just ensured was empty) to the very generic-sounding `lifecycle.containerbuild()` method. That method then calls the very generic-sounding `lxd.Project().execute()` method. Finally, that method checks to see if we're cleaning the pull step with no parts, in which case it blows the container away.

I'd like to see `lxd.Project().execute()` be written more generically. If it was asked to clean with no parts, fine, blow the container away. If it was asked to clean _with_ parts, do nothing. If something else, call `super()`. Then we don't have to let the conditions where the container needs to be removed leak to this line, where we're checking those conditions again. We can simply call `lifecycle.containerbuild('clean')` just like we call `lifecycle.clean()` and expect it to do the right thing.